// @applitools/socket/dist/transports/ws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transport = void 0;
exports.transport = {
    isReady(socket) {
        return socket.readyState === 1; // OPEN
    },
    onReady(socket, callback) {
        socket.addEventListener('open', callback);
        return () => socket.removeEventListener('open', callback);
    },
    onMessage(socket, callback) {
        const handler = (event) => callback(event.data);
        socket.addEventListener('message', handler);
        return () => socket.removeEventListener('message', handler);
    },
    onClose(socket, callback) {
        socket.addEventListener('close', () => callback());
        return () => socket.removeEventListener('close', callback);
    },
    onError(socket, callback) {
        const handler = (event) => callback(event.error);
        socket.addEventListener('error', handler);
        return () => socket.removeEventListener('error', handler);
    },
    send(socket, data) {
        socket.send(data);
    },
};
exports.default = exports.transport;
