// @applitools/eyes-api/dist/Runners.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassicRunner = exports.VisualGridRunner = exports.EyesRunner = void 0;
const NewTestError_1 = require("./errors/NewTestError");
const DiffsFoundError_1 = require("./errors/DiffsFoundError");
const TestFailedError_1 = require("./errors/TestFailedError");
const RunnerOptions_1 = require("./input/RunnerOptions");
const TestResults_1 = require("./output/TestResults");
const TestResultsSummary_1 = require("./output/TestResultsSummary");
const utils = __importStar(require("@applitools/utils"));
class EyesRunner {
    constructor() {
        this._eyes = [];
    }
    /** @internal */
    attach(eyes, spec) {
        if (!this._spec)
            this._spec = spec;
        this._eyes.push(eyes);
    }
    /** @internal */
    async openEyes(options) {
        if (!this._manager)
            this._manager = await this._spec.makeManager(this.config);
        return await this._manager.openEyes(options);
    }
    async getAllTestResults(throwErr = true) {
        var _a;
        if (!this._manager)
            return new TestResultsSummary_1.TestResultsSummaryData();
        const [eyes] = this._eyes;
        const deleteTest = options => this._spec.deleteTest({
            ...options,
            settings: {
                ...options.settings,
                serverUrl: eyes.configuration.serverUrl,
                apiKey: eyes.configuration.apiKey,
                proxy: eyes.configuration.proxy,
            },
        });
        try {
            const summary = await this._manager.getResults({ settings: { throwErr } });
            return new TestResultsSummary_1.TestResultsSummaryData({ summary, deleteTest });
        }
        catch (err) {
            if ((_a = err.info) === null || _a === void 0 ? void 0 : _a.result) {
                const result = new TestResults_1.TestResultsData({ result: err.info.result, deleteTest });
                if (err.reason === 'test failed') {
                    throw new TestFailedError_1.TestFailedError(err.message, result);
                }
                else if (err.reason === 'test different') {
                    throw new DiffsFoundError_1.DiffsFoundError(err.message, result);
                }
                else if (err.reason === 'test new') {
                    throw new NewTestError_1.NewTestError(err.message, result);
                }
            }
            throw err;
        }
    }
}
exports.EyesRunner = EyesRunner;
class VisualGridRunner extends EyesRunner {
    constructor(optionsOrLegacyConcurrency) {
        super();
        if (utils.types.isNumber(optionsOrLegacyConcurrency)) {
            this._legacyConcurrency = optionsOrLegacyConcurrency;
        }
        else if (optionsOrLegacyConcurrency) {
            const options = optionsOrLegacyConcurrency instanceof RunnerOptions_1.RunnerOptionsFluent
                ? optionsOrLegacyConcurrency.toJSON()
                : optionsOrLegacyConcurrency;
            this._testConcurrency = options.testConcurrency;
        }
    }
    /** @internal */
    get config() {
        return {
            type: 'ufg',
            concurrency: this._testConcurrency,
            legacyConcurrency: this._legacyConcurrency,
        };
    }
    get testConcurrency() {
        return this._testConcurrency;
    }
    /** @deprecated */
    get legacyConcurrency() {
        return this._legacyConcurrency;
    }
    /** @deprecated */
    getConcurrentSessions() {
        return this._legacyConcurrency;
    }
}
exports.VisualGridRunner = VisualGridRunner;
class ClassicRunner extends EyesRunner {
    /** @internal */
    get config() {
        return { type: 'classic' };
    }
}
exports.ClassicRunner = ClassicRunner;
