// @applitools/ec-client/dist/commands/start-session.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeStartSession = void 0;
const queue_1 = require("../utils/queue");
const utils = __importStar(require("@applitools/utils"));
const RETRY_BACKOFF = [
    ...Array(5).fill(2000),
    ...Array(4).fill(5000),
    10000, // all next tries with delay 10s
];
function makeStartSession({ settings, req, tunnels }) {
    const queues = new Map();
    return async function createSession({ request, response, logger, }) {
        var _a, _b, _c;
        const requestBody = await utils.streams.toJSON(request);
        logger.log(`Request was intercepted with body:`, requestBody);
        const capabilities = (_b = (_a = requestBody.capabilities) === null || _a === void 0 ? void 0 : _a.alwaysMatch) !== null && _b !== void 0 ? _b : requestBody.desiredCapabilities;
        const options = {
            ...settings.options,
            ...capabilities === null || capabilities === void 0 ? void 0 : capabilities['applitools:options'],
            ...(capabilities &&
                Object.fromEntries(Object.entries(capabilities).map(([key, value]) => [key.replace(/^applitools:/, ''), value]))),
        };
        const session = {
            credentials: { eyesServerUrl: options.eyesServerUrl, apiKey: options.apiKey },
            options,
        };
        if (options.tunnel && tunnels) {
            session.tunnels = await tunnels.acquire(session.credentials);
            session.tunnels.forEach((tunnel, index) => {
                options[`x-tunnel-id-${index}`] = tunnel.tunnelId;
            });
        }
        const applitoolsCapabilities = Object.fromEntries(Object.entries(options).map(([key, value]) => [`applitools:${key}`, value]));
        if (requestBody.capabilities) {
            requestBody.capabilities.alwaysMatch = { ...(_c = requestBody.capabilities) === null || _c === void 0 ? void 0 : _c.alwaysMatch, ...applitoolsCapabilities };
        }
        if (requestBody.desiredCapabilities) {
            requestBody.desiredCapabilities = { ...requestBody.desiredCapabilities, ...applitoolsCapabilities };
        }
        logger.log('Request body has modified:', requestBody);
        const queueKey = JSON.stringify(session.credentials);
        let queue = queues.get(queueKey);
        if (!queue) {
            queue = (0, queue_1.makeQueue)({ logger: logger.extend({ tags: { queue: queueKey } }) });
            queues.set(queueKey, queue);
        }
        request.socket.on('close', () => queue.cancel(task));
        return queue.run(task);
        async function task(signal, attempt = 1) {
            var _a, _b, _c;
            var _d;
            // do not start the task if it is already aborted
            if (signal.aborted)
                return queue.pause;
            const proxyResponse = await req(request.url, {
                body: requestBody,
                io: { request, response, handle: false },
                // TODO uncomment when we can throw different abort reasons for task cancelation and timeout abortion
                // signal,
                logger,
            });
            const responseBody = await proxyResponse.json();
            logger.log(`Response was intercepted with body:`, responseBody);
            if (['CONCURRENCY_LIMIT_REACHED', 'NO_AVAILABLE_DRIVER_POD'].includes((_b = (_a = responseBody.value) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.appliErrorCode)) {
                queue.cork();
                // after query is corked the task might be aborted
                if (signal.aborted)
                    return queue.pause;
                await utils.general.sleep(RETRY_BACKOFF[Math.min(attempt, RETRY_BACKOFF.length - 1)]);
                logger.log(`Attempt (${attempt}) to create session was failed with applitools status code:`, responseBody.value.data.appliErrorCode);
                return task(signal, attempt + 1);
            }
            else {
                queue.uncork();
                if (responseBody.value) {
                    (_c = (_d = responseBody.value).capabilities) !== null && _c !== void 0 ? _c : (_d.capabilities = {});
                    responseBody.value.capabilities['applitools:isECClient'] = true;
                    if (proxyResponse.headers.has('content-length')) {
                        proxyResponse.headers.set('content-length', Buffer.byteLength(JSON.stringify(responseBody)).toString());
                    }
                    session.serverUrl = settings.serverUrl;
                    session.sessionId = responseBody.value.sessionId;
                    session.proxy = settings.proxy;
                    session.capabilities = responseBody.value.capabilities;
                }
                response
                    .writeHead(proxyResponse.status, Object.fromEntries(proxyResponse.headers.entries()))
                    .end(JSON.stringify(responseBody));
                return session;
            }
        }
    };
}
exports.makeStartSession = makeStartSession;
